---
title: "diaQTL Vignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{diaQTL Vignette}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r include=FALSE}
library(knitr)
library(rmarkdown)
knitr::opts_chunk$set(echo = TRUE,collapse=FALSE,comment="##",fig.width=5,fig.height=5)
```

The example data set is a 3x3 half-diallel population in potato. In other words, three founders were crossed in all three pairwise combinations. 

## Structure of the input files
Three input files are needed for QTL analysis: (1) pedigree file, (2) genotype file, (3) phenotype file. In most situations, the PolyOrigin software will be used prior to running diaQTL. The diaQTL function `read_polyancestry` can be used to generate the pedigree and genotype input files from the PolyOrigin output file. 

### 1) Pedigree file
The pedigree file has three columns: id, parent1, and parent2 (maternal effects are not modeled). 
```{r}
pedcsv <- system.file("vignette_data", "potato_ped.csv", package = "diaQTL")
ped <- read.csv(pedcsv, as.is = T)
head(ped)
table(apply(ped[,2:3],1,paste,collapse=" x "))
```

### 2) Genotype file
The first 3 or 4 columns of the genotype file contain the map, with columns for marker name, chromosome, and the position in cM and/or bp (plotting features in the pacakage can use either map position). Subsequent columns contain the genotype probabilities for each individual. 
```{r}
genocsv <- system.file( "vignette_data", "potato_geno.csv", package = "diaQTL" )
geno <- read.csv( genocsv, as.is = T, check.names = F )
geno[1:5,1:4]
```
Genotype probabilities are encoded as strings, following the format exported by the PolyOrigin software:
```{r}
geno[1,10]
```
The integers separated by | on the left side of the equal sign refer to genotype states, and the decimal numbers on the right side of the equal sign are probabilities. Only nonzero probabilities need to be included. There are 100 possible states for F1 populations, and 35 possible states for S1 populations:
```{r}
library( diaQTL )
head( F1codes )
head( S1codes )
```
Each state has four integers, separated by dashes, to indicate which parental chromosomes were inherited. For F1 populations, the maternal chromosomes are labeled 1-4 and the paternal chromosomes 5-8. 

### 3) Phenotype file
In the phenotype input file, the first column should be the individual identifier, followed by columns for different traits, and then optionally any columns with fixed effects to include in the linear model (e.g., block, environment). Only one trait, tuber shape, is provided in the example potato data set.

```{r}
phenocsv <- system.file( "vignette_data", "potato_pheno.csv", package = "diaQTL" )
pheno <- read.csv( phenocsv, as.is = T )
head( pheno )
hist(pheno$tuber_shape,main="",xlab="Tuber shape")
```

To improve normality of the residuals, tuber shape in this data set is defined as log(L/W - 1), where L/W is the average length/width ratio of tubers weighing 6-10 ounces (170-285g). 

## Read the data
After installing and attaching the package, use `read_data` to read in all three files. (If there are fixed effects in the phenotype input file, they need to be specified as well; consult the reference manual.) The parameter `dominance = 2` is used to indicate that only digenic dominance will be modeled. (Higher order dominance effects can be included if desired; consult the manual). By default, markers with the same map position in cM (using whatever numerical precision is present in the input map) are binned to reduce the computing time. The parameter `n.core = 2` is used to enable multicore processing via forking, but this only works from the command line on Linux or MacOS/UNIX systems.
```{r}
data <- read_data(genofile = genocsv,
                  ploidy = 4,
                  pedfile = pedcsv,
                  phenofile = phenocsv, 
                  dominance = 2,
                  n.core = 2)
```

## Setting parameters
diaQTL uses Bayesian regression methods from R package BGLR, which are based on Markov Chain Monte Carlo (MCMC). The function `set_params` determines the burn-in and total number of iterations using the Raftery and Lewis diagnostic from R package `coda`, based on a 95% probability that the estimate for quantile `q` of the additive effects is within the interval `(q-r,q+r)`. For the genome scan, we have found the results based on `q=0.5,r=0.1` to be adequate. Because MCMC is a stochastic process, the results will not be the same each time.
```{r}
params <- set_params( data, trait = "tuber_shape", q=0.5, r=0.1)
params
```

## Single QTL scan
The `scan1` function performs a single QTL scan by regression of the phenotypes on the genotype probabilities. By default, dominance is not incuded in the scan because it takes longer. However, there may be traits where dominance is needed for QTL detection or to refine the QTL position. It is also possible to restrict the QTL scan to a subset of chromosomes. Consult the reference manual for these options.
```{r}
scan1_example <- scan1( data = data,
                        trait = "tuber_shape",
                        params = params, 
                        n.core = 2)
```

## Detection threshold
The function `LODthresh` can be used to estimate the LOD threshold corresponding to a genome-wide false positive rate of 0.05. This function uses a curve fit to simulated datasets with no QTL, which produces results similar to a permutation test (which is also available via the function `scan1_permute`). The genome size for the potato dataset can be estimated from the genetic map.
```{r }
sum(tapply(data@map$cM,data@map$chrom,max))/100 #Morgans 
LODthresh(genome.size=12.2,num.parents=3,ploidy=4) #threshold
```

## Results
The `scan1_summary` function returns the marker with the highest LOD score on each chromosome and a plot of the LOD profile. 
```{r}
ans1 <- scan1_summary(scan1_example,thresh=6.8,position="bp")
ans1$peaks
ans1$plot
```

The results show there is a QTL at 49 Mb on chromosome 10 that explains $R^{2} = 34\%$ of the variance. This matches the location of the classical round locus in potato, which was recently identified as the gene *StOVP20* (Wu et al. 2018; doi.org/10.1038/s41467-018-07216-8). The most significant marker is solcap_snp_c2_25522, which can be passed to the `fitQTL` function to get the estimated haplotype effects, as both a table (`effects`) and barplot (`plots`). Use the parameter `dominance = 2` to include digenic dominance effects. To estimate a 90% Bayesian credible interval (CI) for the effects, the number of MCMC iterations is increased based on the output from `set_params`.
```{r}
params <- set_params( data, trait = "tuber_shape", q=0.05, r=0.05)
params

fit1 <- fitQTL( data = data, 
                trait = "tuber_shape", 
                params = params, 
                marker = "solcap_snp_c2_25522")

#With dominance
fit2 <- fitQTL( data = data, 
                trait = "tuber_shape", 
                params = params, 
                marker = "solcap_snp_c2_25522",
                dominance = 2)

#Compare DIC for additive vs. dominance model
print(data.frame(Add=round(fit1$deltaDIC,0),Dom=round(fit2$deltaDIC,0)),row.names=F)
```

Because the dominance model has more parameters, a penalized measure of goodness-of-fit is needed to determine whether it is superior to the additive model. `fitQTL` returns the Deviance Information Criterion (DIC) for the QTL model relative to the baseline model (no QTL) as deltaDIC. Lower DIC values are better, so the dominance model is selected in this case. The proportion of variance due to additive effects ($h^2$), as well as for higher order effects when included (e.g., $d^2$ for digenic dominance), is returned in `var`. 
```{r}
fit2$var
fit2$plots$add
fit2$plots$dom
```

The dominance plot shows the digenic effects above the diagonal, and below the diagonal is the sum of the additive and digenic effects. The largest additive effect (in magnitude) corresponds to haplotype W6511-1R.4. The function `haplo_get` can be used to extract the dosage of this haplotype across the population.
```{r}
haplos <- haplo_get( data = data, 
                      marker = "solcap_snp_c2_25522")
hist(haplos[,"W6511-1R.4"],main="",xlab="Dosage")
which(haplos[,"W6511-1R.4"] > 1.8)
```

The result shows there are three individuals with two copies of the W6511-1R.4 haplotype, which is possible due to "double reduction." This occurs when a quadrivalent forms in meiosis I and sister chromatid fragments migrate to the same pole in meiosis II. The function `haplo_plot` can be used to visualize the pattern of recombination between parental haplotypes.
```{r}
haplo_plot( data = data, 
            id = "W15268-53R", 
            chrom = 10,
            position = "bp",
            marker = "solcap_snp_c2_25522" )
```

The dark blue segment indicates two copies of the W6511-1R.4 haplotype, and the dashed vertical line shows the position of the QTL at 49 Mb.

## Multiple QTL mapping
In the absence of epistasis, multiple QTL on different chromosomes can be detected adequately with `scan1`. If there is epistasis, or if the QTL are on the same chromosome, other methods are needed. At present, the only option available in `diaQTL` is to conduct a second analysis with `scan1` using the most significant marker from the first scan as a cofactor. Consult the reference manual for more details.
