---
title: "diaQTL Vignette: Potato Example"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{diaQTL Tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r include=FALSE}
library(knitr)
library(rmarkdown)
knitr::opts_chunk$set(echo = TRUE,collapse=FALSE,comment="##",fig.width=5,fig.height=5)
```

## diaQTL Vignette: Potato Example

## Structure of the input files
The example data set is a 3x3 half-diallel population in potato. In other words, three founders were crossed in all three pairwise combinations. Three input files are needed to run the QTL analysis: (1) pedigree file, (2) genotype file, (3) phenotype file. 

### 1) Pedigree file
The pedigree file has four columns: id, population, mother, and father. In the linear model, the mean of each population is estimated as a fixed effect. This allows control over whether reciprocal crosses are assumed to have the same mean (label as one population) or different means (label as separate populations).
```{r}
pedcsv = system.file("vignette_data", "potato_ped.csv", package = "diaQTL")
ped <- read.csv(pedcsv, as.is = T)
head(ped)
table(ped$population)
```

### 2) Genotype file
The first three columns of the genotype file contain the marker name, chromosome, and position in either cM or bp. Subsequent columns contain the genotype probabilities for each individual. The position column in the potato dataset is cM.
```{r}
genocsv = system.file( "vignette_data", "potato_geno.csv", package = "diaQTL" )
geno <- read.csv( genocsv, as.is = T, check.names = F )
geno[1:5,1:3]
```
Genotype probabilities are provided in string format:
```{r}
geno[1,10]
```
The integers separated by | on the left side of the equal sign refer to genotype states, and the decimal numbers on the right side of the equal sign are the probabilities. Only nonzero probabilities need to be included. There are 100 possible states for F1 populations, and 35 possible states for S1 populations:
```{r}
library( diaQTL )
cat("F1 codes\n")
head( F1codes )
cat("\nS1 codes\n")
head( S1codes )
```
Each state has four integers, separated by dashes, to indicate which parental chromosomes were inherited. For F1 populations, the maternal chromosomes are labeled 1-4 and the paternal chromosomes 5-8. 

### 3) Phenotype file
In the phenotype input file, the first column should be the individual identifier, followed by columns for different traits, and then optionally any columns with fixed effects to include in the linear model (e.g., block, environment). Only one trait, tuber shape, is provided in the example potato data set.

```{r}
phenocsv <- system.file( "vignette_data", "potato_pheno.csv", package = "diaQTL" )
pheno <- read.csv( phenocsv, as.is = T )
head( pheno )
```
Tuber shape in this data set is defined as -log(L/W - 1), where L/W is the average length/width ratio of tubers weighing 6-10 ounces (170-285g). Lower values correspond to more elongated tubers.

## Read the data
After installing and attaching the package, use `read_data` to read in all three files. If there are fixed effects (e.g., Environment) in the phenotype input file, they need to be specified at this point (see reference manual).
```{r eval=FALSE}
diallel_example <- read_data( genofile = genocsv,
                              ploidy = 4,
                              pedfile = pedcsv,
                              phenofile = phenocsv)
```
```{r eval=TRUE, echo=FALSE}
load("vignette_data.rda")
```

## Setting parameters
A Bayesian regression model is used in this package, and the posterior distributions are estimated using Markov Chain Monte Carlo (MCMC). The function `set_params` is used to determine the total number of iterations and the number that should be discarded as burn-in. Consult the reference manual on how to change the tolerance parameters if needed. Because MCMC is a stochastic process, the results will not be the same each time.
```{r}
params <- set_params( diallel_example, trait = "tuber_shape" )
params
```

### Single QTL scan
The `scan1` function performs a single QTL scan by regression of the phenotypes on the genotype probabilities. The `scan1_summary` function returns the marker with the highest LOD score on each chromosome and a plot of the LOD profile. By default, dominance is not incuded in the scan because it takes longer. However, there may be traits where dominance is needed for QTL detection, or dominance could lead to refinement of the QTL position. It is also possible to restrict the QTL scan to a subset of chromosomes, or to use multiple cores for faster execution on Linux or UNIX (including MacOS) systems. See the reference manual for these options.
```{r eval=FALSE, echo=TRUE}
scan1_example <- scan1( data = diallel_example,
                      trait = "tuber_shape",
                      params = params )
```

```{r }
ans1 <- scan1_summary(scan1_example)
ans1$peaks
ans1$plot

```

The results show there is a QTL on chromosome 10 that explains $R^{2} = 38\%$ of the variance. The column in the summary table labeled deltaDIC is the Deviance Information Criterion (DIC) relative to the null hypothesis of no QTL. DIC is a penalized measure of model fit (similar to AIC), with lower values indicating better fit. 

## Detection threshold
The function `LODthresh` can be used to estimate the LOD threshold corresponding to a genome-wide false positive rate of 0.05. This function uses a curve fit to simulated datasets with no QTL (Amadeu et al. 2020). The genome size for the potato dataset can be estimated from the input map.
```{r }
potato.genome <- sum(tapply(geno$position,geno$chromosome,max))/100 #Morgans 
threshold <- LODthresh(genome.size=potato.genome,num.parents=3,ploidy=4)
cat(threshold)  #print threshold to standard output
```
Another option is to perform a permutation test with `scan1_permute`, which is computationally intensive.
```{r eval=FALSE, echo=TRUE}
scan1_permute_example <- scan1_permute( data = diallel_example, 
                                   trait = "tuber_shape", 
                                   n.permute = 1000, 
                                   params = params, 
                                   n.core = 4 )
```
The result from `scan1_permute` is a data frame with the LOD and DIC results for each permutation, which can be used to determine the threshold for $\alpha = 0.05$:
```{r eval=TRUE}
alpha <- 0.05
quantile( scan1_permute_example$LOD, probs = 1-alpha )
```
The result is very similar to the threshold from the `LODthresh` function.

## Allele effects
To get the estimated parental allele effects for the QTL on chromosome 10, use the `fitQTL` function at the marker with the highest LOD score. The function returns estimates for $R^{2}$, $deltaDIC$, and QTL $h^2$, as well as the parental allele effects as both a table (`effectsA`) and barplot (`plotA`). 
```{r}
fit1 <- fitQTL( data = diallel_example, 
                       trait = "tuber_shape", 
                       params = params, 
                       marker = "solcap_snp_c2_25522")
fit1$plotA
```
The confidence interval can be reduced by increasing params.

```{r}
fit1 <- fitQTL( data = diallel_example, 
                       trait = "tuber_shape", 
                       params = params, 
                       marker = "solcap_snp_c2_25522")
fit1$plotA
```


To include digenic dominance, use the option `dominance = TRUE`
```{r}
fit_dom <- fitQTL( data = diallel_example,
                   trait = "tuber_shape",
                   params = params_dom,
                   marker = "solcap_snp_c2_25522",
                   dominance = T)

fit_dom$plotD
fit_dom$deltaDIC - fit_default$deltaDIC #better fit with dominance
```

In addition to the `fit1` model, now there is the portion of the QTL heritability explained by the digenic dominance effects and its standard error (`d2$Mean` and `d2$SE`). It also brings means and standard errors for the digenic dominance effects for each allele pair considering all the parents (`effectsD`) and its graphical visualization (`plotD`) in a heat map graphic.

In this example, including dominance increased $R^2$ while lowering the DIC.

You can also verify specific haplotype probabilities (dosage) for any marker for a given individual with:
```{r}
plot_dosage( data = diallel_example, 
             indiv = "W15263-8R", 
             chrom = 10,
             marker = "solcap_snp_c2_25522" )
```

To see the dosage for an specific marker and/or individual:
```{r, eval=FALSE}
dosage( data = diallel_example, 
        indiv = "W15263-8R")

dosage( data = diallel_example, 
        marker = "solcap_snp_c2_25522")
```

## To include a locus as a cofactor in the model

The user can provide a locus name as a cofactor in the model. For it, it should use the `cofactor` option informing a string with the marker name. Here, we performed a new round of analysis considering the `solcap_snp_c2_25522` locus as a cofactor. It takes longer than the default model. 

```{r, eval = FALSE}
ans_default_cof1 <- scan1( data = diallel_example,
                           trait = "tuber_shape",
                           params = params,
                           cofactor = "solcap_snp_c2_25522")

```

To see the output summary:

```{r, eval = FALSE}
scan1_summary(ans_default_cof1) #output not shown
```