#' Find interval to differentiate haplotypes
#' 
#' Finds the smallest interval such that the phased SNP haplotypes are all unique 
#' 
#' The input file (diaQTL_parents.csv) should be generated by \code{\link{read_polyancestry}}. Results can be visualized using \code{\link{phased_parents}}. 
#' 
#' @param filename Name of CSV input file
#' @param marker Name of target marker
#' @param haplotypes Vector of haplotype names to differentiate
#' 
#' @return List containing 
#' \describe{
#' \item{haplo}{Data frame of haplotypes}
#' \item{dendro}{Dendrogram from hclust(method="average")}
#' }
#' 
#' @export
#' @importFrom utils read.csv write.csv
#' @importFrom stats dist as.dendrogram hclust
#' @importFrom labeling extended
#' @importFrom ggfittext geom_fit_text
#' @import ggdendro

haplo_unique <- function(filename,marker,haplotypes) {
  map <- read.csv(filename,as.is=T)
  stopifnot(marker %in% map$marker)
  parents <- colnames(map)[-(1:4)]
  n.parents <- length(parents)
  all.hap <- character(0)
  ploidy <- length(gregexpr(pattern="|",map[1,5],fixed=T)[[1]])+1
  for (i in 1:n.parents) {
    all.hap <- c(all.hap,paste(parents[i],1:ploidy,sep="."))
  }
  stopifnot(haplotypes %in% all.hap)
  
  map <- map[map$chrom==map$chrom[match(marker,map$marker)],]
  m <- nrow(map)
  k <- match(marker,map$marker)
  x <- expand.grid(first=1:k,last=k:m)
  x$size <- x$last-x$first
  x <- x[order(x$size,x$first),]
  x <- x[-1,]
  
  iu <- match(parents,colnames(map))
  if (n.parents==1) {
    tmp <- strsplit(map[,iu],split="|",fixed=T)
    ans <- matrix(as.integer(unlist(tmp)),nrow=m,ncol=ploidy,byrow=T)
  } else {
    tmp <- apply(map[,iu],2,strsplit,split="|",fixed=T)
    ploidy <- length(tmp[[1]][[1]])
    tmp2 <- lapply(tmp,function(z){matrix(as.integer(unlist(z)),ncol=ploidy,nrow=m,byrow=T)})
    ans <- NULL
    for (i in 1:n.parents) {
      ans <- cbind(ans,tmp2[[i]])
    }
  }
  colnames(ans) <- all.hap
  
  finished <- FALSE
  j <- 1
  while (!finished & j <= nrow(x)) {
    first <- x$first[j]
    last <- x$last[j]
    d <- as.matrix(dist(t(ans[first:last,haplotypes]),method="manhattan"))
    diag(d) <- NA
    smallest <- min(apply(d,1,min,na.rm=T))
    
    if (smallest > 0) {
      finished <- TRUE
    } else {
      j <- j + 1
    }
  }
  
  d <- dist(t(ans[first:last,haplotypes]),method="manhattan")
  dhc <- as.dendrogram(hclust(d,method="average"))
  ddat <- dendro_data(dhc,type="rectangle")
  xmax <- max(ddat$segments$yend)
  breaks <- extended(0,xmax,5)
  p <- ggplot(segment(ddat)) + geom_segment(aes(x=y,y=x,xend=yend,yend=xend)) + scale_y_continuous(name="",breaks=NULL,labels=NULL) + geom_fit_text(data=ddat$labels,mapping=aes(xmax=y,xmin=-10,y=x,label=label),place="right") + scale_x_continuous("Hamming Distance",limits=c(-10,xmax),breaks=breaks,labels=breaks) + theme_classic() +  theme(axis.line.y=element_blank(),axis.text.y=element_blank(),axis.ticks.y=element_blank()) 

  if (!finished) {
    print("Reached end of the chromosome and failed to find unique haplotypes")
  }
  return(list(haplo=cbind(map[first:last,1:4],ans[first:last,]),dendro=p))
}
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       